pcf/cli/utils.py

```diff
@@ -0,0 +1,238 @@
+    file_ext = os.path.splitext(filename)[1]
+    basename = os.path.basename(filename)
+
+    if file_ext not in (".json", ".yml", ".yaml"):
+        fail(
+            (
+                "Error: {0} is not a valid PCF config file:\n\nPCF supports JSON and "
+                "YAML config files. Valid file extensions are .json, .yml, and .yaml"
+            ).format(basename)
+        )
+
+    if basename == "pcf.json":
+        for default_config_file in ("pcf.json", "pcf.yml", "pcf.yaml"):
+            if os.path.isfile(default_config_file):
+                return read_config_file(default_config_file)
+        fail(
+            (
+                "Error: could not find a PCF config file.\n\n"
+                "If your config file is not named 'pcf.json', 'pcf.yml', or 'pcf.yaml',\n"
+                "make sure you are using the '--file' option to specify the name of your "
+                "config file, e.g.\n\n\tpcf apply --file my_pcf_config.json"
+            )
+        )
+
+    if os.path.isfile(filename):
+        return read_config_file(filename)
+    else:
+        fail(
+            (
+                "Error: could not find PCF config file {0}\n\n"
+                "If your config file is not named 'pcf.json', 'pcf.yml', or 'pcf.yaml',\n"
+                "make sure you are using the '--file' option to specify the name of your"
+                "config file, e.g.\n\n\tpcf apply --file {0}"
+            ).format(filename)
+        )
+
+
+def read_config_file(filename):
+    """ Loads the JSON/YAML filename and returns it as a dict.
+    """
+    file_ext = os.path.splitext(filename)[1]
+    basename = os.path.basename(filename)
+
+    with open(filename, "r") as config_file:
+        try:
+            if file_ext == ".json":
+                pcf_config = json.load(config_file)
+            else:
+                pcf_config = yaml.safe_load(config_file.read())
+
+            if isinstance(pcf_config, dict):
+                pcf_config = [pcf_config]
+
+            return pcf_config, filename
+
+        except (ValueError, yaml.YAMLError) as error:
+            fail("Error reading PCF config file {0}:\n\n{1}".format(filename, error))
+
+
+def pkg_submodules(package, recursive=True):
+    """ Return a list of all submodules in a given package, recursively by default """
+
+    if isinstance(package, str):
+        try:
+            package = importlib.import_module(package)
+        except ImportError:
+            return []
+
+    submodules = []
+    for _loader, name, is_pkg in pkgutil.walk_packages(package.__path__):
+        full_name = package.__name__ + "." + name
+
+        try:
+            submodules.append(importlib.import_module(full_name))
+        except ImportError:
+            continue
+
+        if recursive and is_pkg:
+            submodules += pkg_submodules(full_name)
+
+    return submodules
+
+
+def particle_class_from_flavor(flavor):
+    """ Return the class object of the given flavor (or None) by searching
```

Comment:
would this work if the user has custom defined flavors not in the core... maybe env var with location or something in the config to specify.

Reply:
Not this function. I wasn't aware of a formal process for building custom Particles external to the package. Is this documented anywhere?

Reply:
https://capitalone.github.io/Particle-Cloud-Framework/docs/build/html/sections/quickstartdeveloper.html

you would use the same way, but would import from a different path

Reply:
That doc recommends pull requests into this package to integrate custom-built Particles and Quasiparticles. Are you suggesting that the CLI should support an optional field in the config file per particle deployment in which users have to specify Python module path strings for this function to look for like this:

```yaml
pcf_name: my_custom_particle
flavor: my_custom_particle_flavor
external_package: my_module.my_custom_particle
aws_resources:
   ...
```

And then the CLI strip out the external_package field before passing the config to the Particle constructor if it finds the right class in `my_module.my_custom_particle`?